### ISSUE TITLE
Issues with Unicode chars

### ISSUE THREAD

**Issue by maoo** [symphonyoss/bot-unfurl#1]
When deploying and testing the bot, I see that some chars don't show up correcty:

18 Oct 2016 1:27:11 pm Testname: Google - http://google.com  http://Google.es/ permite acceder a la informaci�n mundial en castellano, catal�n, gallego, euskara e ingl�s.

**Comment by maoo** [comment]
<img width="777" alt="screen shot 2016-10-18 at 12 08 55" src="https://cloud.githubusercontent.com/assets/327285/19482318/06c40bce-9551-11e6-82f0-1663ba608ad3.png">

**Comment by pmonks** [comment]
Right now the code uses [standard XML escaping (via Apache Commons)](https://github.com/pmonks/bot-unfurl/blob/master/src/bot_unfurl/core.clj#L79) to escape content within the unfurled MessageML message.  It looks like this is not actually the full picture, so I'll have to try to figure out which things need to be escaped in MessageML, and which don't (even though such characters might need to be escaped in real XML).

tl;dr - MessageML is a mess...

**Comment by bruceskingle** [comment]
I can't really tell what the actual problem is from what I can see here, but

a) MessageML is NOT XML
b) MessageML is defined to use UTF-8 encoding

I'm not really sure how to respond to your final comment above.

@maoo what MessageML are you sending which is resulting in the mangled output above?

**Comment by maoo** [comment]
@bruceskingle What I send to the bot is just the text to the url (ie http://google.com), as shown in the screenshot above; the bot responds in Spanish because of my location, I guess.

I believe the issue is related with this code - https://github.com/pmonks/bot-unfurl/blob/master/src/bot_unfurl/core.clj#L79 - as pointed by @pmonks , which is how the bot generates the message text and then escapes it.

**Comment by bruceskingle** [comment]
I'm afraid I don't understand that code, if you can show me the MessageML being generated I may be able to help you.

**Comment by pmonks** [comment]
@maoo if you set logging to debug, the bot will dump raw MessageML messages to the log, which you can then provide to @bruceskingle (via this ticket - please use code fences for readability).  If you've used the [sample log4j file](https://github.com/pmonks/bot-unfurl/blob/master/log4j.xml.sample), you can simply set the `root` priority to `DEBUG`.

**Comment by maoo** [comment]
I've tried again today, but http://google.com fails (see issue #2).

However, when I try to unfurl another url, say `http://adamo.es`, all goes fine and non ASCII chars are properly visualised (as opposed to the initial reported issue).

As requested by @pmonks , this is the messageML being sent:

```
bot-unfurl_1  | 2016-10-24 12:39:16 INFO  core:0 - Found 1 url(s) in message 1w857OqFUqa60_UBeXRmJ3___qgJTEmpdA - posting unfurl message to stream QoCAenzku36x-qlbgPXDbn___qjVgTIpdA
bot-unfurl_1  | 2016-10-24 12:39:16 DEBUG core:0 - Unfurl message: <messageML><b>Fibra optica 1.000Mb</b> - <a href="https://www.adamo.es/"/><br/><i>Servicios de Internet Fibra Óptica. La conexión más rápida de España. Hasta 1000 megas de descarga. Cobertura en España.</i><br/></messageML>
```

<img width="825" alt="screen shot 2016-10-24 at 14 39 43" src="https://cloud.githubusercontent.com/assets/327285/19645979/ba0cd4e0-99f7-11e6-9d3b-57743a3dfcdc.png">

**Comment by bruceskingle** [comment]
As far as I can see there is nothing wrong with this request.

If I hit https://www.adamo.es then I get this response (this is just the first few lines):

```
<!DOCTYPE html>
<html lang="es-ES" prefix="og: http://ogp.me/ns#">
  <head>
          <meta charset="UTF-8">
      <meta name="viewport" content="width=device-width, initial-scale=1.0">
      <meta name="author" content="Adamo Telecom">
```

The document is encoded in UTF-8. If I hit https://www.google.com then I get this:

```
<!doctype html>
<html itemscope="" itemtype="http://schema.org/WebPage" lang="en">
   <head>
      <meta content="Search the world's information, including webpages, images, videos and more. Google has many special features to help you find exactly what you're looking for." name="description">
      <meta content="noodp" name="robots">
      <meta content="/logos/doodles/2016/antoni-van-leeuwenhoeks-384th-birthday-5116176595156992-hp.gif" itemprop="image">
      <link href="/images/branding/product/ico/googleg_lodp.ico" rel="shortcut icon">
      <meta content="Antoni van Leeuwenhoek’s 384th birthday! #GoogleDoodle" property="go:description">
      <meta content="http://www.google.com/logos/doodles/2016/antoni-van-leeuwenhoeks-384th-birthday-5116176595156992-thp.png" property="go:image">
      <meta content="450" property="go:image:width">
      <meta content="200" property="go:image:height">
      <meta content="http://www.google.com/logos/doodles/2016/antoni-van-leeuwenhoeks-384th-birthday-5116176595156992-hp2x.gif" property="go:url">
      <meta content="video.other" property="go:type">
      <meta content="origin" id="mref" name="referrer">
      <title>Google</title>
```

As you can see there is no character set encoding defined which probably means that the output from that page is encoded in **ISO-8859-1** and NOT **UTF-8**.

To get non-ASCII characters through the Symphony API you **MUST** encode them in **UTF-8**.

Note that there are no HTML entities (e.g. &a-acute;) or XML escapes of any kind in the MessageML being sent, it is all in the character set encoding.

**Comment by pmonks** [comment]
@bruceskingle the encoding of the web page has no bearing on the encoding used by the bot for the MessageML message, since the message is constructed independently by the bot.  That message is Unicode up until the point that it's submitted to the Symphony API, since the bot runs on the JVM, which uses Unicode throughout (though _not_ UTF-8 specifically - the JVM is more sophisticated than that and may switch between a variety of encodings internally to optimise memory consumption).

Furthermore, given the evidence @maoo posts above, it appears that the [symphony-java-client](https://github.com/symphonyoss/symphony-java-client) (which this bot uses for interfacing with the Symphony APIs) is also Unicode capable, and is indeed transmitting the message to the server using the correct encoding.  This is unsurprising, given that the symphony-java-client also runs on the same JVM, and therefore has exactly the same level of Unicode support as the bot itself.

@maoo - I believe our pod was upgraded to v1.42.5 in the last couple of days - is it possible that this was a bug in earlier versions of Symphony, and that the upgrade fixed it?

**Comment by bruceskingle** [comment]
@pmonks I beg to differ. 

@maoo has told us that "I've tried again today, but http://google.com fails (see issue #2)." so I don't think that this can be explained by "a bug in earlier versions of Symphony, and that the upgrade fixed it"

If the web page does not specify an encoding then it is quite likely that it is not encoded in UTF-8 because although the default for Java is UTF-8 the W3C default was ISO-8859-1 for HTML4 and earlier.

If the page is encoded in some scheme other than UTF-8 then you will read a sequence of bytes from the server, which in the encoding the server is using makes, for example, a lower case a acute as in "catalán".

Your code or the Java API bindings, I don't know which, is reading that stream of bytes and interpreting it as some encoding. Since the web page does not specify an encoding there is no reliable way for that code to correctly interpret the character data, it just has a stream of bytes of unknown encoding. It seems pretty clear that whatever is happening, this does not include converting the received data into UTF-8.

When that sequence of bytes reaches Symphony it is stored verbatim in the database and then displayed by the Symphony UI which specifies that the data is UTF-8 encoded, but it is not. The hapless browser then cannot display the a acute correctly because is not encoded correctly, it therefore shows the little question mark character which you can see in @maoo 's screen shot at the top of this thread.

If you are seeing those question mark icons then there is no doubt that there are incorrectly encoded extended characters in the message.

**Comment by maoo** [comment]
@pmonks - version 1.42.5 was deployed on the production pod, this weekend there haven't been any deployments on our Dev Pod that have been notified to us; all tests that I've mentioned on this issue refer to foundation-dev.symphony.com, which is a Development Pod.

That said, since the code haven't changed for the last 6 days (at least, not [my fork](https://github.com/maoo/bot-unfurl/commits/master), it seems that the issue was fixed server-side.

**Comment by pmonks** [comment]
@bruceskingle if you look at the response headers upon HTTP GETting http://www.google.com/, you'll notice that the `Content-Type` header is present, and tells the bot what encoding is used (as a parameter on the MIME Type, which is a standard HTTP mechanism).  The bot uses the [clj-http](https://github.com/dakrone/clj-http) HTTP client library, which in turn relies on the [Apache HttpComponents](http://hc.apache.org/) HTTP client library, which respects the `Content-Type` header when reading the stream.

That, coupled with @maoo's observation that the symptom he saw went away by itself sometime in the last 6 days, without any changes to the code, proves beyond reasonable doubt that there was some kind of server issue in Symphony that has since been addressed.

Closed as invalid.
